File: justmeet_backend/src/main/java/com/diegodiome/justmeet_backend/controller/EventController.java
package com.diegodiome.justmeet_backend.controller;

import java.util.List;
import java.util.concurrent.ExecutionException;
import java.util.logging.Level;
import java.util.logging.Logger;

import com.diegodiome.justmeet_backend.model.Category;
import com.diegodiome.justmeet_backend.model.Event;
import com.diegodiome.justmeet_backend.model.EventReporting;
import com.diegodiome.justmeet_backend.service.FirestoreEventService;
import com.diegodiome.justmeet_backend.util.ApiConstants;
import com.diegodiome.justmeet_backend.util.MessageConstants;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.bind.annotation.GetMapping;

@RestController
@RequestMapping(value = ApiConstants.API_EVENTS)
public class EventController {

    @Autowired
    private FirestoreEventService firestoreService;

    private static final Logger LOGGER = Logger.getLogger(FirestoreEventService.class.getName());

    @PostMapping(value = ApiConstants.API_ADD_EVENT, consumes = "application/json")
    public void creteEvent(@RequestBody Event event) {
        try {
            firestoreService.create(event);
            LOGGER.log(Level.WARNING, MessageConstants.EVENT_CREATED_MESSAGE);
        } catch (Exception e) {
            LOGGER.log(Level.WARNING, MessageConstants.EVENT_CREATED_ERROR_MESSAGE);
            e.printStackTrace();
        }
    }

    @DeleteMapping(value = ApiConstants.API_REMOVE_EVENT)
    public void removeEvent(@PathVariable("eventId") String eventId) {
        try {
            firestoreService.remove(eventId);
            LOGGER.log(Level.WARNING, MessageConstants.EVENT_CORRECT_REMOVED_MESSAGE);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @GetMapping(value = ApiConstants.API_GET_EVENT_BY_ID, produces = "application/json")
    public Event getEventById(@PathVariable("eventId") String eventId) {
        try {
            return firestoreService.get(eventId);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    @GetMapping(value = ApiConstants.API_GET_ALL_EVENTS, produces = "application/json")
    public List<Event> getEvents() {
        try {
            return firestoreService.getAll();
        } catch (InterruptedException | ExecutionException e) {
            e.printStackTrace();
        }
        return null;
    }

    @PutMapping(value = ApiConstants.API_UPDATE_EVENT, consumes = "application/json")
    public void updateEvent(@RequestBody Event event) {
        try {
            firestoreService.update(event);
            LOGGER.log(Level.WARNING, MessageConstants.EVENT_CORRECT_UPDATED_MESSAGE);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @GetMapping(value = ApiConstants.API_GET_EVENT_BY_CATEGORY, produces = "application/json")
    public List<Event> getEventsByCategory(@PathVariable("eventCategory") Category category) {
        try {
            return firestoreService.getAllBy(category);
        } catch (InterruptedException | ExecutionException e) {
            e.printStackTrace();
        }
        return null;
    }

    @PostMapping(value = ApiConstants.API_JOIN_EVENT)
    public void enterEvent(@PathVariable("eventId") String eventId, @Value("displayName") String displayName) {
        try {
            firestoreService.joinEvent(displayName, eventId);
            LOGGER.log(Level.INFO, MessageConstants.EVENT_CORRECT_JOINED);
        } catch (InterruptedException | ExecutionException e) {
            e.printStackTrace();
        }
    }

    @PostMapping(value = ApiConstants.API_REPORTING_EVENT, consumes = "application/json")
    public void reportEvent(@RequestBody EventReporting eventReporting) {
        try {
            firestoreService.addReporting(eventReporting);
            LOGGER.log(Level.INFO, MessageConstants.EVENT_COMMENT_CREATED);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @GetMapping(value = ApiConstants.API_GET_EVENT_AVERAGE_RATE, produces = "application/json")
    public double getEventAverageRate(@PathVariable("eventId") String eventId) {
        try {
            return firestoreService.getAverageRate(eventId);
        } catch (InterruptedException | ExecutionException e) {
            e.printStackTrace();
        }
        return 0;
    }
}

File: justmeet_backend/src/main/kotlin/com/diegodiome/justmeet_backend/controller/EventController.kt
package com.diegodiome.justmeet_backend.controller

import com.diegodiome.justmeet_backend.config.constants.ApiConstants.ADD_COM_API
import com.diegodiome.justmeet_backend.config.constants.ApiConstants.ADD_EVENT_API
import com.diegodiome.justmeet_backend.config.constants.ApiConstants.ADD_PART_API
import com.diegodiome.justmeet_backend.config.constants.ApiConstants.ADD_RATE_API
import com.diegodiome.justmeet_backend.config.constants.ApiConstants.ADD_REP_API
import com.diegodiome.justmeet_backend.config.constants.ApiConstants.ADD_REQ_API
import com.diegodiome.justmeet_backend.config.constants.ApiConstants.DEL_EVENT_API
import com.diegodiome.justmeet_backend.config.constants.ApiConstants.EVENT_API
import com.diegodiome.justmeet_backend.config.constants.ApiConstants.GET_COMS_API
import com.diegodiome.justmeet_backend.config.constants.ApiConstants.GET_EVENTS_API
import com.diegodiome.justmeet_backend.config.constants.ApiConstants.GET_EVENT_API
import com.diegodiome.justmeet_backend.config.constants.ApiConstants.GET_REPS_API
import com.diegodiome.justmeet_backend.config.constants.ApiConstants.UPD_EVENT_API
import com.diegodiome.justmeet_backend.model.*
import com.diegodiome.justmeet_backend.repository.EventRepository
import com.diegodiome.justmeet_backend.util.SecurityUtils
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping(value = [EVENT_API])
class EventController  {

    private val eventControllerLogger = LoggerFactory.getLogger(EventController::class.java)

    @Autowired
    lateinit var eventRepository: EventRepository

    @PostMapping(value = [ADD_EVENT_API])
    fun createEvent(@RequestBody newEvent: Event) {
        eventRepository.addElement(newEvent)
    }

    @GetMapping(value = [GET_EVENTS_API])
    fun getEvents() : List<Event> {
        return eventRepository.getEvents()
    }

    @GetMapping(value = [GET_EVENT_API])
    fun getEvent(@PathVariable("eventId") eventId: String) : Event? {
        return eventRepository.getElement(eventId)
    }

    @PutMapping(value = [UPD_EVENT_API])
    fun updateEvent(@RequestBody eventUpdated: Event) {
        eventRepository.updateElement(eventUpdated)
    }

    @DeleteMapping(value = [DEL_EVENT_API])
    fun deleteEvent(@PathVariable("eventId") eventId: String) {
        eventRepository.removeElement(eventId)
    }

    @PostMapping(value = [ADD_COM_API])
    fun addComment(@RequestHeader("Authorization") token: String,
                   @PathVariable("eventId") eventId: String,
                   @RequestBody newComment: Comment) {
        if(SecurityUtils().checkToken(token, newComment.commentCreatorId)) {
            eventRepository.addComment(eventId, newComment)
            return
        }
        eventControllerLogger.warn("[+] User with id : " + newComment.commentCreator + " try to add comment without permission")
    }

    @GetMapping(value = [GET_COMS_API])
    fun getComments(@PathVariable("eventId") eventId: String) : List<Comment> {
        return eventRepository.getComments(eventId)
    }

    @PutMapping(value = [ADD_PART_API])
    fun addParticipant(@PathVariable("eventId") eventId: String,
                       @PathVariable("userId") userId: String) {
        eventRepository.addParticipant(eventId, userId)
    }

    @PostMapping(value = [ADD_REP_API])
    fun addReporting(@RequestBody newReporting: EventReporting) {
        eventRepository.addReporting(newReporting.eventId, newReporting)
    }

    @PutMapping(value = [ADD_RATE_API])
    fun addRate(@PathVariable("eventId") eventId: String,
                @PathVariable("rate") rate: Double) {
        eventRepository.addRate(eventId, rate)
    }

    @GetMapping(value = [GET_REPS_API])
    fun getReports(@PathVariable("eventId") eventId: String) : List<EventReporting> {
        return eventRepository.getReports(eventId)
    }

    @PutMapping(value = [ADD_REQ_API])
    fun addRequest(@PathVariable("eventId") eventId: String
                   , @PathVariable("userId") userId: String) {
        eventRepository.addRequest(eventId, userId)
    }
}